File: service/resource-container/examples/ActivitySensor/src/ActivitySensor.cpp
//******************************************************************
//
// Copyright 2015 Jake Martin Choi All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

/**
 * This file contains the exported symbol.
 */
#include "ActivitySensor.h"

#include <iostream>
#include <stdlib.h>
#include "SysTimer.h"

#ifdef __ANDROID__
#include "OCAndroid.h"
#endif

using namespace ActivitySensorName;


ActivitySensor::ActivitySensor()
{
    heartrate = "";
    activityresult = "";	
    gps = "";
    acceleration = "";
    /*m_weight = "";
    m_height = "";
    m_BMIResult = "";*/
    time(&m_timepstampW);
    time(&m_timepstampH);
}

ActivitySensor::~ActivitySensor()
{

}

int ActivitySensor::executeActivitySensorLogic(std::map<std::string, std::string> *pInputData, std::string *pOutput)
{
	ActivityResult result;

	if(pInputData->find("heartrate") != pInputData->end())
	{
		heartrate = pInputData->at("heartrate");
		time(&m_timepstampW);
	}
	
	if(pInputData->find("acceleration") != pInputData->end())
		heartrate = pInputData->at("acceleration");
	if(pInputData->find("gps") != pInputData->end())
		heartrate = pInputData->at("gps");



	if ((result = makeActivity()) != SUCCESS)
		return -1;

	(*pOutput) = activityresult;

	return 0;
}

ActivityResult ActivitySensor::makeActivity(void)
{
	double heart, timediffsecond;
	double accel, points[3];

	int ActivityResult;

	if(!heartrate.empty() && !acceleration.empty() && !gps.empty())
	{
		heart = std::stod(heartrate);	
		accel = std::stod(acceleration);
		
		timediffsecond = abs(difftime(m_timepstampW, m_timepstampH));

		if(timediffsecond > DIFFTIME)
		{
			ActivityResult = -1;
			std::cout << "[ActivitySensor] : Activity Result Wrong: " << ActivityResult << std::endl;
		}
		else if ((heart > 100) && (accel > 10))
		{
			ActivityResult = (int)EXERCISING; 
			std::cout << "[ActivitySensor] : Activityresult:" << ActivityResult << " EXERCISING" << std::endl;
		}
		else if ((heart < 50) || ((heart > 100) && (accel < 10))) {
			ActivityResult = (int)ABNORMAL;
			
			std::cout << "[ActivitySensor] : Activityresult:" << ActivityResult << " ABNORMAL" << std::endl;
		}
		else
		{
			ActivityResult = (int)NORMAL;
			std::cout << "[ActivitySensor] : Activityresult:" << ActivityResult << " NORMAL" << std::endl;
		}
			

    std::cout << std::endl;

	activityresult  = std::to_string(ActivityResult);

        return SUCCESS;
    }

        else
        {
		heart = -1;
		ActivityResult = -1;
            	std::cout << "[ActivitySensor] :   UNKNOWN: " << ActivityResult << std::endl;
        }

    
    return ERROR;

}

File: service/resource-container/examples/ActivitySensor/ActivityJava/mobile/src/main/java/org/iotivity/service/sample/activitysensor/ActivitySensor.java
//******************************************************************
////
//// Copyright 2015 Susang Park and Jake Choi (Seoul National University) All Rights Reserved.
////
////-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
////
//// Licensed under the Apache License, Version 2.0 (the "License");
//// you may not use this file except in compliance with the License.
//// You may obtain a copy of the License at
////
////      http://www.apache.org/licenses/LICENSE-2.0
////
//// Unless required by applicable law or agreed to in writing, software
//// distributed under the License is distributed on an "AS IS" BASIS,
//// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//// See the License for the specific language governing permissions and
//// limitations under the License.
////
////-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

package org.iotivity.service.sample.activitysensor;

import android.content.Context;

import org.iotivity.service.resourcecontainer.AndroidBundleSoftSensorResource;
import org.iotivity.service.resourcecontainer.RcsResourceAttributes;
import org.iotivity.service.resourcecontainer.RcsValue;

import java.util.Vector;

public class ActivitySensor extends AndroidBundleSoftSensorResource {
    private static final String TAG = "ActivitySensor";
    private static final int HIGH_HEART_RATE = 100;
    private static final int LOW_HEART_RATE = 50;
    private static final int THRESHOLD_ACCELERATION = 100;

    private static final int NORMAL = 1;
    private static final int EXERCISE = 2;
    private static final int ABNORMAL = 3;


    public ActivitySensor(Context context) {
        super(context);
        this.setResourceType("oic.r.activitysensor");
        this.setName("activitySensor");
    }

    @Override
    protected void initAttributes() {
        this.m_attributes.put("heartrate", 0);
        this.m_attributes.put("acceleration", 0);
        this.m_attributes.put("latitude", 0);
        this.m_attributes.put("longitude", 0);
    }

    @Override
    protected void onUpdatedInputResource(String attributeName, Vector<RcsValue> values) {
        double sum = 0, avg = 0;
        int idx = 0;

        for (RcsValue val : values) {
            sum += val.asDouble();
            idx++;
        }
        avg = sum / idx;

        m_mapInputData.put(attributeName, new RcsValue(avg));
    }

    @Override
    protected void executeLogic() {
        // Use sensors for fire detection
        // TODO: Implement proper method to refelect sensors for fire detection. Currently, mechanism is based on simple sensing results combination

        if (m_mapInputData.get("heartrate") == null || m_mapInputData.get("acceleration") == null)
            return;

        double heartrate, acceleration;
        int activity_level = 0;

        heartrate = m_mapInputData.get("heartrate").asDouble();
        acceleration = m_mapInputData.get("acceleration").asDouble();

        if(heartrate > LOW_HEART_RATE && heartrate < HIGH_HEART_RATE){
            activity_level = NORMAL;
        }else if(heartrate > HIGH_HEART_RATE && acceleration > THRESHOLD_ACCELERATION){
            activity_level = EXERCISE;
        }else{
            activity_level = ABNORMAL;
        }

        this.setAttribute("activity-level", new RcsValue(activity_level));
    }

    @Override
    public void handleSetAttributesRequest(RcsResourceAttributes attrs) {
        this.setAttributes(attrs);
        executeLogic();
    }

    @Override
    public RcsResourceAttributes handleGetAttributesRequest() {
        return this.getAttributes();
    }

}
